operators/src/opencl/cl_program.rs

```diff
@@ -0,0 +1,578 @@
+        }
+
+        for (idx, raster) in self.output_rasters.iter().enumerate() {
+            s += &format!(
+                "typedef {} OUT_TYPE{};\n",
+                Self::raster_data_type_to_cl(*raster),
+                idx
+            );
+        }
+
+        s
+    }
+
+    pub fn compile(self, source: &str, kernel_name: &str) -> Result<CompiledCLProgram> {
+        // TODO: check feature inputs
+        ensure!(
+            (self.iteration_type == IterationType::Vector && false)
+                || (self.iteration_type == IterationType::Raster
+                    && !self.input_rasters.is_empty()
+                    && !self.output_rasters.is_empty()),
+            error::CLInvalidInputsForIterationType
+        );
+
+        let typedefs = self.create_type_definitions();
+
+        // TODO: add raster meta data info (NODATA etc.)
+
+        // TODO: add code for pixel to world
+
+        let ctx = ContextBuilder::new().build().context(error::OCL)?; // TODO: make configurable
+
+        let program = ProgramBuilder::new()
+            .src(typedefs)
+            .src(source)
+            .build(&ctx)
+            .context(error::OCL)?;
+
+        let platform = Platform::default(); // TODO: make configurable
+        let device = Device::first(platform).context(error::OCL)?; // TODO: make configurable
+
+        let queue = Queue::new(&ctx, device, None).context(error::OCL)?;
+
+        let mut kernel = KernelBuilder::new();
+        kernel.queue(queue).program(&program).name(kernel_name);
+
+        for (idx, raster) in self.input_rasters.iter().enumerate() {
+            Self::add_buffer_placeholder(&mut kernel, format!("IN{}", idx), *raster);
+        }
+
+        for (idx, raster) in self.output_rasters.iter().enumerate() {
+            Self::add_buffer_placeholder(&mut kernel, format!("OUT{}", idx), *raster);
+        }
+
+        kernel.build().context(error::OCL)?;
+
+        // TODO: raster meta data
+
+        // TODO: feature collections
+
+        let kernel = kernel.build().context(error::OCL)?;
+
+        Ok(CompiledCLProgram::new(
+            kernel,
+            self.iteration_type,
+            self.input_rasters,
+            self.output_rasters,
+        ))
+    }
+}
+
+enum OutputBuffer {
+    U8(Buffer<u8>),
+    U16(Buffer<u16>),
+    U32(Buffer<u32>),
+    U64(Buffer<u64>),
+    I8(Buffer<i8>),
+    I16(Buffer<i16>),
+    I32(Buffer<i32>),
+    I64(Buffer<i64>),
+    F32(Buffer<f32>),
+    F64(Buffer<f64>),
+}
+
+pub struct CLProgramParameters<'a> {
+    input_raster_types: Vec<RasterDataType>,
+    output_raster_types: Vec<RasterDataType>,
+    input_rasters: Vec<Option<&'a TypedRaster2D>>,
+    output_rasters: Vec<Option<&'a mut TypedRaster2D>>,
+}
+
+impl<'a> CLProgramParameters<'a> {
+    fn new(
+        input_raster_types: Vec<RasterDataType>,
+        output_raster_types: Vec<RasterDataType>,
+    ) -> Self {
+        let mut v = Vec::new();
+        v.resize_with(output_raster_types.len(), || None);
```

Comment:
somehow it doesn't work like
> error[E0277]: the trait bound `&mut geoengine_datatypes::raster::typed_raster::TypedRasterNDim<geoengine_datatypes::raster::Dim<[usize; 2]>>: std::clone::Clone` is not satisfied
    --> operators/src/opencl/cl_program.rs:286:29
     |
286  |             output_rasters: vec![None; output_raster_types.len()],
     |                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `std::clone::Clone` is not implemented for `&mut geoengine_datatypes::raster::typed_raster::TypedRasterNDim<geoengine_datatypes::raster::Dim<[usize; 2]>>`

not sure why it can't clone `None`...

Reply:
Your Option would have to be clonable.

Reply:
Option of &mut can't be clone I guess